system/lib/wasmfs/syscalls.cpp

```diff
@@ -17,50 +17,48 @@
 #include <stdarg.h>
 #include <stdlib.h>
 #include <sys/stat.h>
+#include <unistd.h>
 #include <utility>
 #include <vector>
 #include <wasi/api.h>
-#include <unistd.h>
 
 extern "C" {
 
 using namespace wasmfs;
 
 // Copy the file specified by the pathname into JS.
 // Return a pointer to the JS buffer in HEAPU8.
+// Caller must free the returned pointer.
 void* emscripten_wasmfs_read_file(char* path) {
   struct stat file;
   int err = 0;
   err = stat(path, &file);
   if (err < 0) {
-    emscripten_console_error(
-      "Fatal error in FS.readFile");
+    emscripten_console_error("Fatal error in FS.readFile");
     abort();
   }
-  
-  size_t size = file.st_size;
+
+  // file.st_size could exceed uint32_t in wasm64.
+  assert(file.st_size < std::numeric_limits<uint32_t>::max());
+  uint32_t size = file.st_size;
   uint8_t* result = (uint8_t*)malloc((size + sizeof(size)));
```

Comment:
It looks like this is allocating a structure of some kind: one field of size "size" and one field of size "sizeof(size)". Is that correct? If so, docs explaining the format would be useful, as this needs to match on the reader side.

Reply:
Mmm I can add some more comments around this. This was done to avoid having to return the size to `FS.readfile()` via another C++ function. Instead we can return the length and the actual buffer contents in one function. The caller should be made aware of this however.